File: 2326-Spiral-Matrix-IV/neet.py
class Solution:
    def spiralMatrix(self, m: int, n: int, head: Optional[ListNode]):
        left, right = 0, n
        top, bottom = 0, m
        res = [[-1] * n for _ in range(m)]

        while head:
            # 1. left to right
            for i in range(left, right):
                if not head: return res
                res[top][i] = head.val
                head = head.next
            top += 1

            # 2. top to bottom
            for i in range(top, bottom):
                if not head: return res
                res[i][right-1] = head.val
                head = head.next
            right -= 1

            # 3. right to left
            for i in range(right - 1, left - 1, -1):
                if not head: return res
                res[bottom-1][i] = head.val
                head = head.next
            bottom -= 1

            # 4. bottom to top
            for i in range(bottom - 1, top - 1, -1):
                if not head: return res
                res[i][left] = head.val
                head = head.next
            left += 1

        return res

File: 2326-Spiral-Matrix-IV/neet.js
var spiralMatrix = function(m, n, head) {
    let left = 0, right = n;
    let top = 0, bottom = m;
    const res = Array.from({length: m}).map(() => Array(n).fill(-1));

    while (head) {
        // from left to right
        for (let i = left; i < right; i++) {
            if (!head) return res;
            res[top][i] = head.val;
            head = head.next;
        }
        top += 1

        // from top to bottom
        for (let i = top; i < bottom; i++) {
            if (!head) return res;
            res[i][right - 1] = head.val;
            head = head.next;
        }
        right -= 1;

        // from right to left
        for (let i = right - 1; i > left - 1; i--) {
            if (!head) return res;
            res[bottom - 1][i] = head.val;
            head = head.next;
        }
        bottom -= 1;

        // from bottom to top
        for (let i = bottom - 1; i > top - 1; i--) {
            if (!head) return res;
            res[i][left] = head.val;
            head = head.next;
        }
        left += 1;
    }

    return res;
};